faunadb-java-dsl/src/main/java/com/faunadb/client/types/Types.java

```diff
@@ -0,0 +1,231 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.errors.FaunaException;
+import org.joda.time.Instant;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.lang.String.format;
+
+/**
+ * Auxiliary class used to create types representation while decoding user defined objects.
+ */
+public final class Types {
```

Comment:
This is interesting. I think it's interface is important but, I feel like it'd be better placed in the `Value` class so we could do things like:

```java
value.asMapOf(String.class, User.class);
// Or
value.asCollectionOf(User.class);
```

Also, the `SimpleType` looks like a re-implementation of the codecs map. Shouldn't we try to re-use the codecs?